zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -137,15 +163,76 @@ impl<F: Field> ExecutionGadget<F> for BeginTxGadget<F> {
         );
 
         // Transfer value from caller to callee
+        // Use cb.query_word as TransferWithGasFeeGadget
+        // expects words instead of expressions for tx_value and gas_fee
+        let effective_tx_value = cb.query_word();
+        let effective_gas_fee = cb.query_word();
+        cb.condition(is_tx_invalid.expr(), |cb| {
+            cb.require_equal(
+                "effective_tx_value == 0",
+                effective_tx_value.expr(),
+                true.expr(),
+            );
+            cb.require_equal(
+                "effective_gas_fee == 0",
+                effective_gas_fee.expr(),
+                true.expr(),
+            );
+        });
+
+        cb.condition(not::expr(is_tx_invalid.expr()), |cb| {
+            cb.require_equal(
+                "effective_tx_value == tx_value",
+                effective_tx_value.expr(),
+                tx_value.expr(),
+            );
+            cb.require_equal(
+                "effective_gas_fee == gas_fee",
+                effective_gas_fee.expr(),
+                gas_fee.product().expr(),
+            );
+        });
+
         let transfer_with_gas_fee = TransferWithGasFeeGadget::construct(
             cb,
             tx_caller_address.expr(),
             tx_callee_address.expr(),
-            tx_value.clone(),
-            mul_gas_fee_by_gas.product().clone(),
+            effective_tx_value.clone(),
+            effective_gas_fee.clone(),
             &mut reversion_info,
         );
 
+        // Verify transfer
+        let sender_balance_prev = transfer_with_gas_fee.sender.balance_prev();
+
+        let add_tx_value_and_mul_gas_fee_by_gas = cb.query_word();
+
+        // tx_value.int_value + gas_fee.int_value
+        let total_eth_cost = AddWordsGadget::construct(
+            cb,
+            [tx_value.clone(), gas_fee.product().clone()],
+            add_tx_value_and_mul_gas_fee_by_gas.clone(),
+        );
+
+        // balance_not_enough = sender_balance_prev.int_value < tx_value.int_value +
+        // gas_fee.int_value
+        let balance_not_enough = LtWordGadget::construct(
```

Comment:
- Give a bitter comment like `// Check if the account ETH balance is sufficient`.
- Remove new lines here so it's easier to see that all of these statements are related